### PR TITLE
Add SeriesDict, SeriesList, and all related changes

### PR DESCRIPTION
**Pull request by thijs-obj**
## Changes
* add `instance_dtype` to Series. For 'normal' series this is the class dtype. For structural types this contains a structure that describes what the dtypes of the parts of the values are. e.g. for a `SeriesList` the `instance_dtype` might be `['int64']` to describe a list of integers.
* add `SeriesDict`
  * BigQuery only
  * `STRUCT` in SQL
* add `SeriesList`
  * BigQuery only
  * `ARRAY` in SQL
* Use `Series.supported_db_dtype` to track which databases are supported by each Series subtype.

## Open Issues
* Support slices in SeriesList: https://github.com/objectiv/objectiv-analytics/issues/704
* Make this work with concat(): https://github.com/objectiv/objectiv-analytics/issues/703
* Make this work with merge(): https://github.com/objectiv/objectiv-analytics/issues/705
* TODO: There might be more work related to other functions when we start to use this more.

Related issues:
* Support SeriesJson https://github.com/objectiv/objectiv-analytics/issues/702

### CODE REVIEW

File: bach/bach/operations/concat.py [hunk 1]
```diff
@@ -302,8 +302,9 @@ def _get_concatenated_object(self) -> Series:
             expression=Expression.column_reference(final_result_series.name),
             index={} if self.ignore_index else main_series.index,
             group_by=None,
-            index_sorting=[] if self.ignore_index else main_series.index_sorting,
             sorted_ascending=None,
+            index_sorting=[] if self.ignore_index else main_series.index_sorting,
+            instance_dtype=series_cls.dtype  # TODO: make this work for structural types too
```
Comment (thijs-obj) [inline bach/bach/operations/concat.py:307]: https://github.com/objectiv/objectiv-analytics/issues/703

File: bach/bach/series/series_list.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+                    .copy_override(expression=expression)
+            elif isinstance(sub_dtype, dict):
+                from bach import SeriesDict
+                return self._series \
+                    .copy_override_type(SeriesDict, instance_dtype=sub_dtype) \
+                    .copy_override(expression=expression)
+            else:
+                raise Exception(f'Unexpected type of sub_dtype. sub_dtype: {sub_dtype}')
+
+        elif isinstance(key, slice):
```
Comment (thijs-obj) [inline bach/bach/series/series_list.py:205]: https://github.com/objectiv/objectiv-analytics/issues/704

File: bach/bach/types.py [hunk 3]
```diff
@@ -209,24 +280,45 @@ def get_series_type_from_dtype(self, dtype: DtypeOrAlias) -> Type['Series']:
         Given a dtype string or a dtype alias, will return the correct Series object to represent such data.
         """
         self._real_init()
+        # lists and dicts are hardcoded exceptions for now.
+        # TODO: make this nicer, e.g. make sure that no class can register this as an alias
```
Comment (thijs-obj) [inline bach/bach/types.py:284]: I think we can leave the TODOs here until we get to the next database that we want to support. Should work okayish for now

File: bach/bach/types_bq.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    if pos != len(tokens) - 1:
+        raise ValueError(f'Unexpected tokens after last parsed tokens. '
+                         f'Last parsed tokens position: {pos}, total tokens: {len(tokens)}')
+    return result
+
+
+# Code below is not super clean, and we should have more tests. But we have tests for bq_db_dtype_to_dtype()
+# so we can always refactor this in the future. For now it will do fine.
+
+def _tokenize(bq_db_dtype: str):
```
Comment (thijs-obj) [inline bach/bach/types_bq.py:35]: This is based on Tom's original code found here: https://github.com/objectiv/objectiv-analytics/pull/474/files#diff-1cd9b5d39d7a622aa1860ad34fa66ffe9866228c02df08eb9be1eb64653969f1

File: bach/bach/types.py [hunk 5]
```diff
@@ -39,34 +40,99 @@
 """
 
 
-DtypeOrAlias = Union[Type, str]
+Dtype = str
+StructuredDtype = Union[List[Any], Mapping[str, Any], Dtype]
```
Comment (KathiaBarahona) [inline bach/bach/types.py:44]: I think I'm a bit confused with this two types. Not sure if it's the naming, but I keep going back to this file to know what a function using this typing is expecting

File: bach/bach/series/series.py [hunk 6]
```diff
@@ -298,6 +320,11 @@ def expression(self) -> Expression:
         """ INTERNAL: Get the expression"""
         return self._expression
 
+    @property
+    def instance_dtype(self) -> StructuredDtype:
+        """ Get the instance_dtype """
+        return deepcopy(self._instance_dtype)
+
```
Comment (KathiaBarahona) [inline bach/bach/series/series.py:327]: I think it's worth it to extend a bit the explanation here and highlight what is the difference between `Series.dtype` and `Series.instance_dtype`.

File: bach/tests/functional/bach/test_series_list.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        name='int_list',
+        dtype=['int64']
+    )
+    df['str_list'] = SeriesList.from_value(
+        base=df,
+        value=['a', 'b', 'c'],
+        name='str_list',
+        dtype=['string']
+    )
+    print(df.dtypes)
```
Comment (KathiaBarahona) [inline bach/tests/functional/bach/test_series_list.py:42]:
```suggestion
```

File: bach/bach/series/series.py [hunk 8]
```diff
@@ -321,59 +349,100 @@ def get_class_instance(
...
+        return cls.supported_value_to_literal(dialect=dialect, value=value, dtype=dtype)
+
+    @classmethod
+    def value_to_expression(
+            cls,
+            dialect: Dialect,
+            value: Optional[Any],
+            dtype: StructuredDtype
+    ) -> Expression:
```
Comment (KathiaBarahona) [inline bach/bach/series/series.py:407]: maybe? I'm aware you might need to do some changes due to invariant/covariance warnings
```suggestion
    ) -> ConstValueExpression:
```

File: bach/bach/types.py (hunk 5 above)
Comment (thijs-obj) [reply bach/bach/types.py:44]: Does perhaps the name `InstanceDtype` make more sense to you? Or do you have a specific suggestion for naming/documentation/something else?

File: bach/bach/series/series_list.py [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+            if isinstance(sub_dtype, Dtype):
+                new_dtype = sub_dtype
+                return self._series \
+                    .copy_override_dtype(dtype=new_dtype) \
+                    .copy_override(expression=expression)
+            elif isinstance(sub_dtype, list):
+                return self._series \
+                    .copy_override(instance_dtype=sub_dtype) \
+                    .copy_override(expression=expression)
```
Comment (KathiaBarahona) [inline bach/bach/series/series_list.py:196]: isn't it the same?
```suggestion
                return self._series \
                    .copy_override(instance_dtype=sub_dtype, expression=expression)
```

File: bach/bach/series/series.py (hunk 6 above)
Comment (thijs-obj) [reply bach/bach/series/series.py]: I added more documentation in a commit that I just pushed: https://github.com/objectiv/objectiv-analytics/pull/700/commits/3bc15427a29a24ec7ecb735648c306b6ebd5a64d

let me know what you think

File: bach/bach/series/series_list.py (hunk 9 above)
Comment (thijs-obj) [reply bach/bach/series/series_list.py:206]: It is the same. Wrote it like this such that all three if/elif clauses looked visually similar, hoped that would make it easier to read. Will change it

File: bach/bach/series/series.py (hunk 8 above)
Comment (thijs-obj) [reply bach/bach/series/series.py:411]: ok. Changed List to Sequence in relevant places

File: bach/bach/series/series_dict.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+    such a Series must then be dictionaries with the exact same field names as the instance-dtype, and the
+    specified data types for the fields.
+
+    | Example instance_dtype
+    | BigQuery db_dtype: ``'STRUCT<a INT64, b ARRAY[FLOAT64], c <STRUCT x BOOL, y: INT64>>'``
+    | instance_dtype: ``{'a': 'int64', b: ['float64'], 'c': {'x': 'bool', 'y': 'int64'}``
+
+    .. note::
+        SeriesDict is only supported on BigQuery.
+        On Postgres use SeriesJson for similar functionality.
```
Comment (jansentom) [inline bach/bach/series/series_dict.py:37]: We should probably wrap this somewhere in the future?

**Review by jansentom — APPROVED** [review]
Very nice way to initially support this. With two DBs "supported" this will work quite well. Some things need a bit more thoughtful generalisation later on though (how to deal with types that are not supported but could be wrapped for example), but nice to NOT have this in this PR.

**Comment by thijs-obj** [comment]
Discussed out-of-band: reviewers okay with merging this

**Pull request merged**